### PR TITLE
[SPARK-28208][BUILD][SQL] Upgrade to ORC 1.5.6 including closing the ORC readers

### PR DESCRIPTION
**Pull request by omalley**
## What changes were proposed in this pull request?

It upgrades ORC from 1.5.5 to 1.5.6 and adds closes the ORC readers when they aren't used to 
create RecordReaders.

## How was this patch tested?

The changed unit tests were run.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @omalley !

File: pom.xml [hunk 1]
```diff
@@ -139,7 +139,7 @@
     <kafka.version>2.3.0</kafka.version>
     <derby.version>10.12.1.1</derby.version>
     <parquet.version>1.10.1</parquet.version>
-    <orc.version>1.5.5</orc.version>
```
Comment (dongjoon-hyun) [inline pom.xml]: Could you do the following to update the dependency file?
```
$ dev/test-dependencies.sh --replace-manifest
```

**Comment by dongjoon-hyun** [comment]
Retest this please

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
All the UT failures are due to ORC 1.5.6 regression which was reported at ORC 1.5.6 RC1 voting (ORC-525).

**Comment by dongjoon-hyun** [comment]
I'll try to make a patch for the missing part and send to your branch, `omalley:spark-28208`, during weekend, @omalley . Thank you again.

**Comment by dongjoon-hyun** [comment]
Interesting. It passed with `-Phadoop-3.2`.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala [hunk 2]
```diff
@@ -184,6 +184,7 @@ class OrcFileFormat
 
       val requestedColIdsOrEmptyFile = OrcUtils.requestedColumnIds(
         isCaseSensitive, dataSchema, requiredSchema, reader, conf)
+      reader.close()
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala]: Can we do:

```scala
      val requestedColIdsOrEmptyFile = {
        Utils.tryWithResource(OrcFile.createReader(filePath, readerOptions)) { reader =>
          OrcUtils.requestedColumnIds(
            isCaseSensitive, dataSchema, requiredSchema, reader, conf)
        }
      }
```

or, `try` and `finally`?

?

**Comment by dongjoon-hyun** [comment]
Hi, @omalley . I made a PR to your branch. Could you review and merge that to pass the UT successfully?
- https://github.com/omalley/spark/pull/1

That will make this PR pass the following suites.
- FileBasedDataSourceSuite
- OrcQuerySuite.
- OrcV1QuerySuite.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @omalley .

**Review by felixcheung — COMMENTED** [review]
btw, this isn't a `[BUILD]` PR, it's `[SQL]`

**Comment by kiszk** [comment]
Gentle ping, @omalley

**Comment by srowen** [comment]
Ping @omalley to update or close

**Comment by omalley** [comment]
Ok, sorry for the delay. I just got back to this.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you so much, @omalley .
Thank you, @HyukjinKwon , @felixcheung , @kiszk , @srowen .

This changes inevitable to give more controllability to close the ORC files. If there occurs ORC test suite failures due to ORC file leaks, we can fix them now and improve Spark code base.

Merged to `master`.

**Comment by HyukjinKwon** [comment]
https://github.com/apache/spark/pull/25006#discussion_r298788206 is a valid comment. Can we fix it too?

**Comment by dongjoon-hyun** [comment]
Sure. I'll make a follow-up, @HyukjinKwon ~

**Comment by HyukjinKwon** [comment]
nope, that should be fine enough.

**Pull request closed without merging**